src/devices/src/virtio/block/request.rs

```diff
@@ -214,37 +214,40 @@ impl Request {
             .map_err(ExecuteError::Seek)?;
 
         match self.request_type {
-            RequestType::In => {
-                mem.read_from(self.data_addr, diskfile, self.data_len as usize)
-                    .map_err(ExecuteError::Read)?;
-                METRICS.block.read_bytes.add(self.data_len as usize);
-                METRICS.block.read_count.inc();
-                return Ok(self.data_len);
-            }
-            RequestType::Out => {
-                mem.write_to(self.data_addr, diskfile, self.data_len as usize)
-                    .map_err(ExecuteError::Write)?;
-                METRICS.block.write_bytes.add(self.data_len as usize);
-                METRICS.block.write_count.inc();
-            }
-            RequestType::Flush => match diskfile.flush() {
-                Ok(_) => {
+            RequestType::In => mem
+                .read_exact_from(self.data_addr, diskfile, self.data_len as usize)
+                .map(|_| {
+                    METRICS.block.read_bytes.add(self.data_len as usize);
+                    METRICS.block.read_count.inc();
+                    self.data_len
+                })
+                .map_err(ExecuteError::Read),
+            RequestType::Out => mem
+                .write_all_to(self.data_addr, diskfile, self.data_len as usize)
+                .map(|_| {
+                    METRICS.block.write_bytes.add(self.data_len as usize);
+                    METRICS.block.write_count.inc();
+                    self.data_len
```

Comment:
Hmmm, why are we now returning `self.data_len` for `Write` request? The `len` field of an entry from the used ring should contain the number of bytes written to memory, not of the ones read from there. For`Write` request only the status byte is written to memory.

Reply:
Good catch! Updated.